### PR TITLE
Add Streaming support

### LINKED ISSUE #6: better way to handle batches/streams?
for the Calendar app we need to expand the _busytimes_ (same event might repeat multiple times), so a basic calendar sync might trigger thousands of expansions..

waiting for the whole operation to complete before we display the events to the user is not really a good approach (might take many seconds to finish the sync), so we really need some sort of event listener that allow us to handle busytimes in batches (maybe queue 10-100 busytimes before dispatching an event) and/or a way to handle busytimes as they are expanded (emit one event for each busytime).

we also need a way to know that the whole expansion/sync finished.

I guess SMS and Contacts would also need something similar.

### PR DESCRIPTION
**Pull request by millermedeiros**
I ended up using 2 different APIs (one for the `client` and another one for the `service`).

---

closes #6

### CODE REVIEW

**Comment by millermedeiros** [comment]
# API v1

 `write()`, `error()` and `close()` only exists on the `service`, and `ServiceStream` is **not** an `EventEmitter` (find it cleaner and less error-prone to write `stream.write('foo')` than `stream.emit('data', 'foo')`)

the `ClientStream` inherits from `EventEmitter` and implements the `abort()` method.
- `write()` because that is what [node.js](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback) and  [whatwg streams](https://streams.spec.whatwg.org/#ws-write) uses.
- `abort()` was inspired by [whatwg writable stream](https://streams.spec.whatwg.org/#ws-abort) but just realized that on readable streams it is actually called [`cancel()`](https://streams.spec.whatwg.org/#rs-cancel) so maybe we should rename it on the client.
- `error()` was because node.js streams can emit an `error` event.
- `close()` was because we node.js streams emit a `close` event.

## Client API

``` js
var stream = client.stream('contacts', { some: 'data' });
stream.on('data', onData);
stream.on('error', onError);
stream.on('close', onClose);
stream.abort('because I want').then(doStuff, onError);
```

## Service API

``` js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.onabort = cancelEverything;
    if (data.foo === 123) return stream.error('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

File: lib/client-stream.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+module.exports = ClientStream;
+
+function ClientStream(client, id) {
+  this._client = client;
+  this._id = id;
+}
+
+ClientStream.prototype = {
+  __proto__: Emitter.prototype,
```
Comment (wilsonpage) [inline lib/client-stream.js]: Is this smelly? Can we use `Object.create()`.

File: lib/service.js [hunk 2]
```diff
@@ -114,6 +122,8 @@ function ServicePrivate(service, name) {
   this.contract = null;
   this.methods = {};
   this.channels = {};
+  this.streams = {};
+  this._activeStreams = {};
```
Comment (wilsonpage) [inline lib/service.js]: NIT: `_` not required as this is inside the 'private' class.

Comment (millermedeiros) [inline lib/service.js]: sure

File: lib/client-stream.js (hunk 1 above)
Comment (millermedeiros) [inline lib/client-stream.js]: sure

File: lib/service-stream.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  this._id = id;
+  this._state = 'writable';
+}
+
+ServiceStream.prototype = {
+  abort: function(reason) {
+    this._validateState('abort');
+    this._state = 'aborted';
+    if (!('onabort' in this)) {
+      throw new Error('stream.onabort() not implemented by service.');
```
Comment (wilsonpage) [inline lib/service-stream.js]: What if the service isn't interested in this event?

Comment (wilsonpage) [inline lib/service-stream.js]: An reason why we're not using `Emitter` here?

File: lib/service.js [hunk 4]
```diff
@@ -75,6 +78,11 @@ Service.prototype.method = function(name, fn) {
   return this;
 };
 
+Service.prototype.stream = function(name, fn) {
+  this.private.streams[name] = fn;
```
Comment (wilsonpage) [inline lib/service.js]: Let's make an `this.private.addStream()` to prevent touching internals.

File: lib/service-stream.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
+
+'use strict';
+
+module.exports = ServiceStream;
+
+function ServiceStream(service, id) {
```
Comment (wilsonpage) [inline lib/service-stream.js]: I feel like it might make sense to pass the client's `channel` into the `ServiceStream` so that we can pipe messages down it. Right now it looks like we're broadcasting stream messages to all clients. This might mean we can kill the new `._broadcast` API which is confusing me.

File: lib/service.js [hunk 6]
```diff
@@ -360,13 +396,17 @@ ServicePrivate.prototype.listen = function() {
  */
 ServicePrivate.prototype.broadcast = function(type, data) {
   debug('broadcast', type, data);
+  this._broadcast('broadcast', {
+    type: type,
+    data: data
+  });
+};
+
+ServicePrivate.prototype._broadcast = function(messageId, data) {
```
Comment (wilsonpage) [inline lib/service.js]: As mentioned above: If we passed the `this.channels[client]` into the `ServiceStream` it could send messages directly to the client on it's own.

File: lib/client.js [hunk 7]
```diff
@@ -5,6 +5,7 @@
  * Dependencies
  */
```
Comment (wilsonpage) [inline lib/client.js:7]: We should decide on a 'private' pattern (like in `Service`) and use it across all our classes.

File: lib/service-stream.js (hunk 5 above)
Comment (wilsonpage) [inline lib/service-stream.js]: A stream will only ever have one client right?

File: lib/client.js [hunk 8]
```diff
@@ -242,6 +247,47 @@ Client.prototype.method = function(method) {
...
+    args: args,
+    id: id
+  }).catch(err => {
+    this.onstreamerror({ id: id, data: err });
+  });
+
+  return stream;
+};
+
+Client.prototype.onstreamdata = function(broadcast) {
```
Comment (wilsonpage) [inline lib/client.js]: Might be simpler to remove these three events and replace them with:

``` js
Client.prototype.onstreamevent = function(event) {
  this._activeStreams[event.stream].emit(event.type, event.data);
};
```

Then if you need to do any internal house-keeping on any particular events you can bind to their public API just after they're created.

``` js
stream
  .on('close', deleteStream);
  .on('error', deleteStream);
```

File: lib/service-stream.js (hunk 5 above)
Comment (millermedeiros) [inline lib/service-stream.js]: yes, a stream should only have a single client.

**Comment by millermedeiros** [comment]
# API v2

I just updated the PR with a new API, implementation is still not as clean as it could be but I think the new API is closer to the [WHATWG streams](https://streams.spec.whatwg.org/) and simpler to use.

## Client API

``` js
var stream = client.stream('contacts', { some: 'data' });
var binding = stream.listen(onData);
binding.cancel(); // removes listener
stream.closed.then(onClose, onAbort);
stream.cancel('because I want').then(doStuffAfterCancel, onCancelError);
```

## Service API

``` js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.oncancel = cancelEverything;
    if (data.foo === 123) return stream.abort('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

File: threads.js [hunk 9]
```diff
@@ -164,7 +164,63 @@ ChildThread.prototype.destroyProcess = function() {
...
+    this._emitter.emit('data', data);
+  },
+
+  listen: function(callback) {
+    var emitter = this._emitter;
+    emitter.on('data', callback);
+    return {
+      cancel: function() {
+        emitter.off('data', callback);
+      }
```
Comment (millermedeiros) [inline threads.js]: uhm.. now that I'm thinking about it, it would be *better* to have an `unlisten` or `removeListener` method instead of returning a new object.

File: lib/client-stream.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  },
+
+  listen: function(callback) {
+    var emitter = this._emitter;
+    emitter.on('data', callback);
+    return {
+      cancel: function() {
+        emitter.off('data', callback);
+      }
+    };
```
Comment (millermedeiros) [inline lib/client-stream.js]: uhm.. now that I'm thinking about it, it would be _better_ to have an `unlisten` or `removeListener` method instead of returning a new object.

**Comment by millermedeiros** [comment]
# API v3

closer to the [WHATWG streams](https://streams.spec.whatwg.org/) and simpler to use.

`listen()` was inspired by [dart:async:Stream](https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:async.Stream#id_listen) but I added the `unlisten()` method and accepted a single callback because that makes more sense in our case -- we already have the `closed` promise and we won't have as many features as the dart:async:StreamSubscription.

## Client API

``` js
var stream = client.stream('contacts', { some: 'data' });
stream.listen(onData);
stream.unlisten(onData);
stream.closed.then(onClose, onAbort);
stream.cancel('because I want').then(doStuffAfterCancel).catch(onCancelError);
```

## Service API

``` js
threads.service('my-service')
  .stream('contacts', (stream, data) => {
    stream.cancel = cancelEverything;
    if (data.foo === 123) return stream.abort('invalid argument');
    stream.write('stuff');
    stream.close();
  });
```

and now instead of throwing errors I'm returning rejected promises on `write`, `close` and `abort` (in case service tries to call `write` after `abort/close` or call `abort/close` multiple times in a row) kinda following the whatwg writable stream behavior. -- we ignore the action without throwing an error since we assume that after `abort`/`close` the stream is going to be destroyed and also because throwing errors on async operations is an anti-pattern.

Calling `cancel` on a `service` that doesn't override `stream.cancel` will also return a rejected promise.

**Comment by millermedeiros** [comment]
I believe the patch is ready for review, added unit tests for the edge cases, JSDoc comments for public APIs, refactored... I'll squash the commits after I get r+

File: lib/client.js (hunk 7 above)
Comment (millermedeiros) [inline lib/client.js:7]: on the streams I used `this._ = PrivateStream()` since it's easy to type and it's common to have `_` as the start of private members..

 `this._.foo` also looks cool because `._.` is almost an emoji :D

Comment (millermedeiros) [inline lib/client.js:7]: exposing the private methods/properties is good because it allows us to unit test, see: https://github.com/millermedeiros/threads/commit/843425d5e0a2a64f2eb1794bf7e0eefd3758af5c#diff-3d6cd7d35d59ce33f7b9c33ff2d59a20R20

Comment (wilsonpage) [inline lib/client.js:7]: :heart: ._.

File: lib/service.js [hunk 11]
```diff
@@ -7,6 +7,7 @@
 
 var thread = require('./thread-global');
 var utils = require('./utils');
+var ServiceStream = require('./service-stream');
```
Comment (wilsonpage) [inline lib/service.js]: Move this to line 8 to satisfy my irrational OCD-ness :stuck_out_tongue_winking_eye:

File: lib/service.js [hunk 12]
```diff
@@ -360,13 +407,17 @@ ServicePrivate.prototype.listen = function() {
  */
 ServicePrivate.prototype.broadcast = function(type, data) {
   debug('broadcast', type, data);
+  this.postMessage('broadcast', {
+    type: type,
+    data: data
+  });
+};
+
+ServicePrivate.prototype.postMessage = function(messageId, data) {
```
Comment (wilsonpage) [inline lib/service.js]: Can we get some jsdocs on all private function too? Even though they're not public it helps me follow the code :)

File: test/service-stream.test.js [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    var activeStream;
+    var STREAM_ID = 'fake-stream-id';
+
+    setup(function() {
+      service = threads.service('test-stream');
+    });
+
+    function callStream(methodName, fn) {
+      service.stream(methodName, fn);
+      service.private.channels['123'] = {
```
Comment (wilsonpage) [inline test/service-stream.test.js]: Would it be possible to test these edge cases without tying implementation details to the tests?

Comment (millermedeiros) [inline test/service-stream.test.js]: it would if I could access the `ServiceStream` constructor -- which is a private constructor at the moment and our test infrastructure don't really load the src files, just the built file.

File: lib/client.js [hunk 14]
```diff
@@ -242,6 +245,37 @@ Client.prototype.method = function(method) {
   });
 };
 
+Client.prototype.stream = function(method) {
```
Comment (wilsonpage) [inline lib/client.js:266]: Docs needed

File: threads.js [hunk 15]
```diff
@@ -731,15 +995,56 @@ ServiceInternal.prototype.onrequest = function(request) {
...
+ */
+ServicePrivate.prototype.onstream = function(method, request) {
+  debug('stream', method.name);
+  var fn = this.streams[method.name];
+  if (!fn) throw new Error(ERRORS[4]);
+
+  var id = method.id;
+  var client = request.client;
+  var channel = this.channels[client];
+  var stream = new ServiceStream(id, channel, this.id, client);
```
Comment (wilsonpage) [inline threads.js]: Number of args passsed to `ServiceStream` is getting lengthy. Perhaps a `params` object might be more readable. What do you think?

File: threads.js [hunk 16]
```diff
@@ -575,7 +704,106 @@ ManagerInternal.prototype.createThread = function(descriptor) {
...
+function PrivateServiceStream(target, id, channel, serviceId, clientId) {
+  this.target = target;
+  this.id = id;
+  this.channel = channel;
+  this.client = clientId;
+  this.state = 'writable';
+  this.message = new utils.Messages(this, serviceId, []);
+}
+
+PrivateServiceStream.prototype.validateState = function(actionName, state) {
```
Comment (wilsonpage) [inline threads.js:841]: Why doesn't this function return a promise? Can't see any async stuff.

**Comment by wilsonpage** [comment]
Just a few NITs really. Mainly meed jsdoc coverage to make code easier to understand. Overall this is looking really quality! Your code is slick :)

File: threads.js (hunk 16 above)
Comment (millermedeiros) [inline threads.js:841]: it returns a promise because `cancel`, `abort`, `close` and `write` should only really send data to the client if it's on a valid state and they should return rejected promises when they fail.. so it made sense to do `return this.validateState('cancel', 'canceled').then(...)` but I'll replace the logic since it's confusing.

**Comment by millermedeiros** [comment]
squashed commits and exposed the `ServiceStream` as `threads.service.Stream`
